Net iplist

guiniol opened the pull request:
This PR adds a module to display the list of IP of the computer. IPs can be blacklisted by prefix. By default the loopback is blacklisted.

tobes commented on py3status/modules/net_iplist.py:
```diff
@@ -0,0 +1,90 @@
+        self.separator = ' '
+        self.no_connection = 'no connection'
+        self.ignore_prefix = ['127.']
+        pass
```
These should be defined at the class level rather than in __init__()

tobes commented on py3status/modules/net_iplist.py:
```diff
@@ -0,0 +1,90 @@
+    no_connection: string to display if there are no non-ignored IPs
+        (default: 'no connection')
+    ignore_prefix: list of prefixes to ignore.
+        (default: ['127.'])
```
Hmm,  think I'd be inclined to do ignoring as reg expressions (althought all the `.` make it a pain)  maybe use wild cards so `ignore = ['127.*', '192.168.*.100']` type thing to be honest I've not got too many Ips on my machines to worry about this too much.

guiniol commented:
I'll see what I can do. Adding regexp is easy, but if we want users to type '127.*' and it should be understood as '127\..*', that's something else.

guiniol commented:
Added support for wildcards. The `fnmatch` module already does it.

tobes commented:
Ok, I have the following questions

1) What is the purpose of this module?  Is it purely about showing my local IP addresses?

2) Is there other information that I might later want eg which interface is connected to each IP address?

3) Is this only for IPv4?  what about IPv6?

4) Is there a reason that you are not using something like `ifconfig` or `ip addr show`?

5) Is there a reason for no `format` config parameter?  This is something that we are trying to get all modules to have for consistency etc.

guiniol commented:
> 1) What is the purpose of this module? Is it purely about showing my local IP addresses?

Yes, that's the idea.
 
> 2) Is there other information that I might later want eg which interface is connected to each IP address?

That's actually a good idea since I already have the information. I'll add that.

> 3) Is this only for IPv4? what about IPv6?

IPv4 only for now, I'll look into IPv6

> 4) Is there a reason that you are not using something like ifconfig or ip addr show?

Because I can't be sure ifconfig or ip addr will be installed. This is the only reliable way to find it that I could find. It's what this does too: https://pypi.python.org/pypi/netifaces

> 5) Is there a reason for no format config parameter? This is something that we are trying to get all modules to have for consistency etc.

I have an idea but the format is a bit weird:
1. `format` has the `{iface_list}` placeholder (I don't remember what the actual word for "placeholder" is)
2. `iface_list` is of the form `iface_format``iface_separator``iface_format``iface_separator``iface_format`...
3. `iface_format` is a sub format with the `{iface}` and `{ip_list}` placeholders
4. `ip_list is of the form `ip_format``ip_separator``ip_format``ip_separator``ip_format`...
5. `ip_format` is a sub format with the `{ip}` placeholder
6. `*_separators` are just strings

It's a bit complex, but I don't see how to do it simpler than that.

guiniol commented:
@tobes: new answers now that I rewrote the module

> 1) What is the purpose of this module? Is it purely about showing my local IP addresses?

Still yes.

> 2) Is there other information that I might later want eg which interface is connected to each IP address?

You can do this now (actually, it does it by default)

> 3) Is this only for IPv4? what about IPv6?

Handles both now, and it's easy to choose which to show.

> 4) Is there a reason that you are not using something like ifconfig or ip addr show?

I switch to using `ip addr`. It was the easiest way to handle multiple ipv4 addresses on one interface.

> 5) Is there a reason for no format config parameter? This is something that we are trying to get all modules to have for consistency etc.

Added. I'm not sure if there is a better solution than what I did (`format` + `format_iface`) but I don't really see how else to handle the list of interfaces/IPs in `format`.

tobes commented on py3status/modules/net_iplist.py:
```diff
@@ -0,0 +1,152 @@
+                                                                  'ip': txt_ip,
+                                                                  'ip6': txt_ip6})
+            iface_txt += self.iface_sep
+            iface_txt = iface_txt[:-len(self.iface_sep)]
```
This just removes the text you added, I think you meant to do it outside the loop.

Anyhow personally I'd put all the bits in a list and the join them as that way you don't need to worry about the extra bit on the end

guiniol replied:
> Anyhow personally I'd put all the bits in a list and the join them as that way you don't need to worry about the extra bit on the end

Good thinking.

tobes commented on py3status/modules/net_iplist.py:
```diff
@@ -0,0 +1,152 @@
+            idx += 1
+
+            if iface is None:
+                continue
```
why not do the blacklisting here?  better not collect the data than to have to remove it imho

tobes commented on py3status/modules/net_iplist.py:
```diff
@@ -0,0 +1,152 @@
+    def _get_data(self):
+        iface_re = re.compile(r'\d+: (\w+):')
+        ip_re = re.compile(r'\s+inet (\d{1,3}.\d{1,3}.\d{1,3}.\d{1,3})/')
+        ip6_re = re.compile(r'\s+inet6 ([\da-f:]+)/')
```
Probably not a huge performance deal but if you compiled these in `__init__()` then you would only need to do it once.

Did you consider using groups?  eg `(?P<name>....)`

tobes commented on py3status/modules/net_iplist.py:
```diff
@@ -0,0 +1,152 @@
+    iface_sep = ' '
+    ip_blacklist = []
+    ip_sep = ','
+    no_connection = 'no connection'
```
`format_no_connection` would be more to the py3status naming conventions

tobes commented on py3status/modules/net_iplist.py:
```diff
@@ -0,0 +1,152 @@
+        ip_re = re.compile(r'\s+inet (\d{1,3}.\d{1,3}.\d{1,3}.\d{1,3})/')
+        ip6_re = re.compile(r'\s+inet6 ([\da-f:]+)/')
+
+        txt = check_output(['ip', 'address', 'show']).decode('utf-8').split('\n')
```
`splitlines()` is nicer than `split('\n')` because it understands newlines better but on unix output not such an issue

guiniol replied:
I didn't know about `splitlines`

tobes commented on py3status/modules/net_iplist.py:
```diff
@@ -0,0 +1,152 @@
+
+        idx = 0
+        data = {}
+        while idx < len_txt:
```
why not `for line in txt:`  we know interface name lines don't start with `' '` so this whole loop can probably be reduced in complexity.

guiniol replied:
I'm not sure that's possible. I explicitly move the index because I don't know how many lines there are per interface. For now, the code really shows that we look for an interface, and once we have one, we parse until the next interface. I could rewrite it with a variable to remember the current interface. I'll see if it makes it cleaner/more readable.

tobes commented on py3status/modules/net_iplist.py:
```diff
@@ -0,0 +1,152 @@
+            if ips == {}:
+                to_del.add(iface)
+        for iface in to_del:
+            del data[iface]
```
again if we didn't add this data we wouldn't have to remove it later

tobes commented on py3status/modules/net_iplist.py:
```diff
@@ -0,0 +1,152 @@
+        (default ['lo'])
+    iface_sep: string to write between interfaces.
+        (default ' ')
+    ip_blacklist: list of IPs to ignore. Accepts shell-style wildcards.
```
an example would be good, either here or as an example at the end (see https://github.com/ultrabug/py3status/blob/master/doc/README.md#docstring)

guiniol replied:
Sure.

tobes commented on py3status/modules/net_iplist.py:
```diff
@@ -0,0 +1,152 @@
+    color_good: IPs to show
+
+Requires:
+    iproute2
```
i'd change this to `ip: utility found in iproute2 package`

guiniol replied:
Ok.

guiniol commented:
@tobes Done!

For the `for line in txt:` loop, it (invisibly) assumes that we will match an iface before matching IPs. Otherwise, it'll error.

For the blacklisting, like I said, I'd rather keep it in the main loop.

I also added the option to keep interfaces without IPs, and interfaces can be blacklisted with wildcards too (useful if you want to ignore all `tap*` devices or something).

tobes commented:
@guiniol looking good, I think I'd be inclined to change the `ip` references to `ip4` just to make it slightly clearer.

I find this useful to reduce the number of commits, just update my last one
```
git add -p  # add any changes
git commit --amend   # you can change the commit message here too
git push -f  # need to force the updates
```

I'll do a proper review tomorrow when I have some time.

guiniol commented:
> @guiniol looking good, I think I'd be inclined to change the ip references to ip4 just to make it slightly clearer.

Thanks ^^
Actually, `ip addr` shows `inet` and `inet6` which is why I went with `ip` and `ip6`.

> I find this useful to reduce the number of commits, just update my last one
> 
> git add -p  # add any changes
> git commit --amend   # you can change the commit message here too
> git push -f  # need to force the updates

I actually like lots of small commits, but I do use `--amend` when I have something to change. If you're referencing c52c6ac, It must be merged in ca58e69

> I'll do a proper review tomorrow when I have some time.

Sure. Take your time.

tobes commented on py3status/modules/net_iplist.py:
```diff
@@ -0,0 +1,151 @@
+class Py3status:
+    cache_timeout = 30
+    format = 'Network: {format_iface}'
+    format_iface = '{iface}: v4{{{ip}}} v6{{{ip6}}}'
```
how about `'{iface}:[ {ip}][ {ip6}]'`

One of my interfaces has no IPv6 and it looks funny being blank.  Also I find the `{` an unnecessary distraction.

tobes commented:
> Actually, ip addr shows inet and inet6 which is why I went with ip and ip6.

I think that is because IPv6 was not really in use when ip was first written

guiniol commented on py3status/modules/net_iplist.py:
```diff
@@ -0,0 +1,151 @@
+Display the list of current IPs.
+
+This can exclude specified IPs and displays "no connection" if there are no IPs
+to display.
```
Done.

guiniol commented on py3status/modules/net_iplist.py:
```diff
@@ -0,0 +1,151 @@
+
+    def __init__(self):
+        self.iface_re = re.compile(r'\d+: (?P<iface>\w+):')
+        self.ip_re = re.compile(r'\s+inet (?P<ip>\d{1,3}.\d{1,3}.\d{1,3}.\d{1,3})/')
```
Don't you mean `r'\s+inet (?P<ip>[0-9\.]+)/'` :þ

guiniol commented:
> I think that is because IPv6 was not really in use when ip was first written

I was on the fence for that. Done.

ultrabug commented:
It's about time I start merging @guiniol contributions, can we have an updated labeling on ready for merge ones please ?

How do you feel about them @guiniol ?

guiniol commented:
I think this one is done. @tobes?

ultrabug commented:
Looks very nice, thanks !
